src/core/gui/PdfFloatingToolbox.cpp

```diff
@@ -200,16 +195,22 @@ void PdfFloatingToolbox::createStrokes(PdfMarkerStyle position, PdfMarkerStyle w
         stroke->addPoint(Point(rect.x2, h, -1));
         stroke->setStrokeCapStyle(StrokeCapStyle::BUTT);
 
+        dirtyRange.addPoint(rect.x1, h - 0.5 * w);
+        dirtyRange.addPoint(rect.x2, h + 0.5 * w);
+
         strokes.push_back(stroke);
-        layer->addElement(stroke);
-        page->fireElementChanged(stroke);
     }
 
+    doc->lock();
+    for (Stroke* s: strokes) {
+        layer->addElement(s);
+    }
+    doc->unlock();
+    page->fireRangeChanged(dirtyRange);
```

Comment:
Functionally, I suppose this works because the page is the only `PageListener`, and the only thing `XojPageView::elementChanged` does is rerender the element. But technically, we still need to call `fireElementChanged` because we could potentially have additional `PageListener`s.

Reply:
Yes. I think I'll just add an optional Range parameter to fireElementChanged.

Overall, `Page::fireRangeChanged` does not make a lot of sense to me and should be removed, I guess. Out of scope of this PR ofc.